web/html/src/manager/minion/ansible/ansible-var-yaml-editor.tsx

```diff
@@ -0,0 +1,287 @@
+import React, { useCallback, useEffect, useState } from "react";
+
+import { useFormikContext } from "formik";
+import yaml from "js-yaml";
+import get from "lodash/get";
+
+import { Button, DropdownButton } from "components/buttons";
+import { Field, MultiField } from "components/formik/field";
+import { Form } from "components/formik/Form";
+import { Panel } from "components/panels/Panel";
+import { MessagesContainer } from "components/toastr/toastr";
+
+import BooleanEditor from "./variables/boolean-editor";
+import DictionaryEditor from "./variables/dictionary-editor";
+import ExtraVariabl from "./variables/extra-var";
+import ListEditor from "./variables/list-editor";
+import StringEditor from "./variables/string-editor";
+
+const isDictionary = (obj) => {
+  if (typeof obj !== "object" || obj === null || Array.isArray(obj)) return false;
+  return Object.values(obj).every(
+    (val) => typeof val === "string" || typeof val === "number" || typeof val === "boolean" || val === null
+  );
+};
+
+const variablesList = ["List", "Dictionary", "String", "Boolean"];
+
+type Props = {
+  data: Record<string, any>;
+  onDataChange: (values: Record<string, any>) => void;
+  onExtraVarChange: (extravalues: string) => void;
+};
+
+const AnsibleVarYamlEditor = (props: Props) => {
+  const { data, onDataChange, onExtraVarChange } = props;
+  const [visibleInputPath, setVisibleInputPath] = useState(null);
+  const [varType, setVarType] = useState(null);
+  const [expandAllClicked, setExpandAllClicked] = useState(false);
+  const [expandAllState, setExpandAllState] = useState(false);
+
+  const generateId = (path) => `id_${path.split(".").join("_")}`;
+
+  // Titles for collapse top level
+  const levelOneTitles = (obj) => Object.keys(obj);
+
+  // Nested level
+  const nestedLevelTitles = (prefix, value) => {
+    const current = get(value, prefix) as any;
+    let paths: string[] = [];
+
+    // don't render seconnd level for array or Dictionary
+    if (!current || typeof current !== "object" || Array.isArray(current) || isDictionary(current)) return [];
+
+    for (const key in current) {
+      const path = `${prefix}.${key}`;
+      const val = current[key];
+      paths.push(path);
+
+      if (typeof val === "object" && val !== null && !Array.isArray(val) && !isDictionary(val)) {
+        paths = paths.concat(nestedLevelTitles(path, value));
+      }
+    }
+    // console.log(paths)
+    return paths;
+  };
+
+  const YamlPreview = () => {
+    const { values } = useFormikContext<Record<string, any>>();
```

Comment:
The general recommendation by Formik is to use `Form` instead of `useFormikContext`. Is there a specific reason to use that here? The value listener then moves into the form's `onChange` and everything else should require very little changes.